src/ts/forwarder.ts

```diff
@@ -0,0 +1,188 @@
+  }
+
+  /**
+   * Takes care of replay protection and signs a meta-transaction.
+   * @param data Target contract address, value (wei) to send, and the calldata to exeucte in the target contract
+   */
+  public async signMetaTransaction(data: T) {
+    const forwarderAddr = await this.getForwarderAddress();
+
+    const encodedReplayProtection = await this.replayProtectionAuthority.getEncodedReplayProtection();
+    const encodedCallData = this.getEncodedCallData(data);
+    const encodedMetaTx = this.encodeMetaTransactionToSign(
+      encodedCallData,
+      encodedReplayProtection,
+      this.replayProtectionAuthority.getAddress(),
+      forwarderAddr
+    );
+
+    const signature = await this.signer.signMessage(
+      arrayify(keccak256(encodedMetaTx))
+    );
+
+    const params = this.getForwardParams(
+      forwarderAddr,
+      data,
+      encodedReplayProtection,
+      signature
+    );
+
+    return params;
+  }
+
+  /**
+   * Fetches the forward parameters. Used when signing a new
+   * meta-transaction.
+   * @param to Forwarder address
+   * @param data Target, value and calldata
+   * @param replayProtection Encoded replay protection
+   * @param signature Signature to authorise meta-transaction
+   */
+  protected abstract getForwardParams(
+    to: string,
+    data: T,
+    replayProtection: string,
+    signature: string
+  ): ForwardParams;
+
+  /**
+   * Encodes the forward function and its arguments such that
+   * included in the data field of an Ethereum Transaction.
+   * @param params ForwardParameters
+   */
+  public abstract async encodeSignedMetaTransaction(
+    params: ForwardParams
+  ): Promise<string>;
+
+  /**
+   * Authorises the deployment of a smart contract with a deterministic address
+   * @param initCode Bytecode of contract
+   */
+  /**
+   * Signs a meta-transaction to deploy a contract via CREATE2.
+   * Takes care of replay protection.
+   * @param initCode Bytecode for the smart contract
+   */
+  public async signMetaDeployment(initCode: string) {
+    const forwarderAddr = await this.getForwarderAddress();
+
+    const encodedReplayProtection = await this.replayProtectionAuthority.getEncodedReplayProtection();
+    const encodedMetaTx = this.encodeMetaTransactionToSign(
+      initCode,
+      encodedReplayProtection,
+      this.replayProtectionAuthority.getAddress(),
+      forwarderAddr
+    );
+    const signature = await this.signer.signMessage(
+      arrayify(keccak256(encodedMetaTx))
+    );
+
+    const params: DeploymentParams = {
+      to: forwarderAddr,
+      signer: this.signer.address,
+      initCode,
+      replayProtection: encodedReplayProtection,
+      replayProtectionAuthority: this.replayProtectionAuthority.getAddress(),
+      chainId: this.chainID,
+      signature: signature,
+    };
+
+    return params;
+  }
+
+  /**
+   * The address that will appear in the msg.sender of target contract
+   */
+  public abstract async getForwarderAddress(): Promise<string>;
```

Comment:
This should be renamed to `getAddress` - since we're already on the forwarder class. You may also be able to make this not async if you pass the correct things into each of the constructors.

Reply:
I wanted to make a distinction between the signers address and the forwarder address.

Reply:
I kept this as ForwarderAddress() to avoid confusing if it will fetch signer address or forwarder.

Reply:
I still think it should be changed to just getAddress, it's an instance method on a forwarder object. Eg forwarder.getAddress() would always be expected to return the forwarder address. The signer address is always available via this.signer.address.

Reply:
done